src/mo_frontend/typing.ml

```diff
@@ -1379,6 +1385,24 @@ and check_exp' env0 t exp : T.typ =
       warn env exp.at "M0155" "operator may trap for inferred type%a"
         display_typ_expand t;
     t
+  | ToCandidE (ot, exps), _ ->
+    begin match t with
+    | T.Prim T.Blob -> begin
+        let ts = List.map (infer_exp env) exps in
+        ot := T.Tup ts;
+        T.Prim T.Blob
+      end
+    | _ ->
+      error env exp.at "MXXXX" "to_candid produces Blob, not type%a"
+        display_typ_expand t
+    end
```

Comment:
Instead of matching on the structure of `t` you should just check that `Blob` is subtype of `t` (which could be Any).
Also, you need to check the ts (or ot) all `shared`.